### PR TITLE
New implementation of second algorithm finished ( O(N*log^2(N) )

### PR DESCRIPTION
**Pull request by Blauben**

### CODE REVIEW

**Comment by Blauben** [comment]
For some reason the github tests fail with:
```
/home/runner/work/polyhedral-gravity-model/polyhedral-gravity-model/src/polyhedralGravity/util/UtilityContainer.h:10:10: fatal error: gtest/internal/gtest-param-util.h: No such file or directory
   10 | #include <gtest/internal/gtest-param-util.h>
```

Do you know what could be the cause of that?
@schuhmaj

**Comment by schuhmaj** [comment]
> For some reason the github tests fail with:
> 
> ```
> /home/runner/work/polyhedral-gravity-model/polyhedral-gravity-model/src/polyhedralGravity/util/UtilityContainer.h:10:10: fatal error: gtest/internal/gtest-param-util.h: No such file or directory
>    10 | #include <gtest/internal/gtest-param-util.h>
> ```
> 
> Do you know what could be the cause of that? @schuhmaj

Precisely what's written there 🙃: You're including `<gtest/internal/gtest-param-util.h>` in `UtilityContainer.h` which you shouldn't.

File: src/polyhedralGravity/model/KDTree/KdDefinitions.h [hunk 1]
```diff
@@ -137,4 +135,44 @@ namespace polyhedralGravity {
...
+    struct PlaneEvent {
+        PlaneEventType type;
+        /**
+         * The candidate plane suggested by the face included in this struct.
+         */
+        Plane plane;
+        /**
+         * The index of the face that generated this candidate plane.
+         */
+        unsigned faceIndex;
```
Comment (schuhmaj) [inline src/polyhedralGravity/model/KDTree/KdDefinitions.h:160]: Please use the full type name `unsigned int`

File: src/polyhedralGravity/model/KDTree/KDTree.cpp [hunk 2]
```diff
@@ -23,41 +25,164 @@ namespace polyhedralGravity {
...
+        std::vector<PlaneEvent> events{std::move(generatePlaneEvents(splitParam))};
+        size_t trianglesMin{0}, trianglesMax{splitParam.indexBoundFaces.size()}, trianglesPlanar{0};
+        //traverse all the events
+        int i{0};
+        while (i < events.size()) {
+            //poll a plane to test
+            Plane &candidatePlane = events[i].plane;
+            //for each plane calculate the faces whose vertices lie in the plane. Differentiate between the face starting in the plane, ending in the plane or all vertices lying in the plane
+            size_t p_start{0}, p_end{0}, p_planar{0};
+            //count all faces that end in the plane, this works because the PlaneEvents are sorted by position and then by PlaneEventType
```
Comment (schuhmaj) [inline src/polyhedralGravity/model/KDTree/KDTree.cpp:64]: If the `events` vector is sorted after `PlaneEventType`. How should it be possible that the first `while` is every executed again? At least the condition `i < events.size()` in four `while` loops seems to some degree faulty 🤔.

File: src/polyhedralGravity/model/KDTree/KDTree.cpp [hunk 3]
```diff
@@ -23,41 +25,164 @@ namespace polyhedralGravity {
...
-                if (testedPlaneCoordinates.find(candidatePlane.axisCoordinate) != testedPlaneCoordinates.cend()) {
-                    continue;
+            //clip plane coordinates to voxel
+            const auto [minAxisCoordinate, maxAxisCoordinate] = clipToVoxel(splitParam.boundingBox, splitParam.splitDirection, minPoint, maxPoint);
+            // if the triangle is perpendicular to the split direction, generate a planar event with the candidate plane in which the triangle lies
+            if (minAxisCoordinate == maxAxisCoordinate) {
+                events.emplace_back(
+                        PlaneEventType::planar,
+                        Plane{
+                                .axisCoordinate = minAxisCoordinate,
```
Comment (schuhmaj) [inline src/polyhedralGravity/model/KDTree/KDTree.cpp:113]: Caution ⚠️! This is `C++20`. However, the implementation currently shall be `set(CMAKE_CXX_STANDARD 17)` --> either raise the standard in the `CMakeLists.txt` and keep it or don't use *designated initializers*

File: src/polyhedralGravity/model/KDTree/KDTree.h [hunk 4]
```diff
@@ -8,9 +8,13 @@
 #include <algorithm>
 #include <array>
 #include <memory>
+#include <queue>
+#include <thrust/iterator/transform_iterator.h>
 #include <unordered_set>
 #include <utility>
-#include <thrust/iterator/transform_iterator.h>
+
+#define MIN (0)
```
Comment (schuhmaj) [inline src/polyhedralGravity/model/KDTree/KDTree.h:16]: Just for completeness - Don't use preprocessor macros if you can use type-safe C++ constructs (I know, you've already fixed this one branch further 👍😉)

File: src/polyhedralGravity/model/KDTree/KDTree.cpp [hunk 5]
```diff
@@ -23,41 +25,164 @@ namespace polyhedralGravity {
...
+                            .axisCoordinate = maxAxisCoordinate,
+                            .orientation = splitParam.splitDirection},
+                    index);
+        });
+        //sort the events by plane position and then by PlaneEventType. Refer to {@link PlaneEventType} for the specific order
+        std::sort(events.begin(), events.end());
+        return events;
+    }
+
+    TriangleIndexLists<2> KDTree::generateTriangleSubsets(const std::vector<PlaneEvent> &planeEvents, const Plane &plane, const bool minSide) {
```
Comment (schuhmaj) [inline src/polyhedralGravity/model/KDTree/KDTree.cpp:137]: Is the `std::vector` actually faster than a `std::set`? Did you benchmark it? I refer here to the definitions

File: src/polyhedralGravity/model/KDTree/KDTree.cpp [hunk 6]
```diff
@@ -75,34 +200,44 @@ namespace polyhedralGravity {
...
         }
-        //include equalT in the boxes triangles for further subdivision
-        greaterT->insert(greaterT->cend(), equalT->cbegin(), equalT->cend());
-        return {costUpper, std::array{(std::move(lessT)), (std::move(greaterT))}};
+        //if empty space is cut off, reduce cost by 20%
+        return {factor * costUpper, false};
+    }
+
+    template<typename... Points>
+    std::array<double, sizeof...(Points)> KDTree::clipToVoxel(const Box &box, const Direction direction, Points... points) {
```
Comment (schuhmaj) [inline src/polyhedralGravity/model/KDTree/KDTree.cpp:228]: `template` is a source file? Don't forget to explcitly instantiate it with the types which you utilize.

File: src/polyhedralGravity/model/KDTree/KDTree.h [hunk 7]
```diff
@@ -91,15 +95,14 @@ namespace polyhedralGravity {
...
-                Array3Triplet vertexTriplet =  {
-                    vertices[face[0]],
-                    vertices[face[1]],
-                    vertices[face[2]]
-                } ;
+                const auto &face = faces[faceIndex];
+                Array3Triplet vertexTriplet = {
+                        vertices[face[0]],
+                        vertices[face[1]],
+                        vertices[face[2]]};
```
Comment (schuhmaj) [inline src/polyhedralGravity/model/KDTree/KDTree.h:105]: probably, some issue with our current clang-tidy. I'd prefer the old style here. If you have spare time, you're gldaly invited to find out which option fixes it 😇

File: src/polyhedralGravity/model/KDTree/TreeNodeFactory.h [hunk 8]
```diff
@@ -7,6 +7,8 @@
 
 #include <memory>
 
+#define MAX_RECURSION_DEPTH 64
```
Comment (schuhmaj) [inline src/polyhedralGravity/model/KDTree/TreeNodeFactory.h:10]: Same thing here if not yet fixed: `constexpr int MAX_RECURSION_DEPTH = 64`

**Review by schuhmaj — APPROVED** [review]
Nice 👍 Just some small comments (which might be already fixed given that I was quiet late reviewing this)

⚠️The important thing is the one C++20/ C++17 comment ⬇️!

File: src/polyhedralGravity/model/KDTree/KDTree.h (hunk 4 above)
Comment (Blauben) [reply src/polyhedralGravity/model/KDTree/KDTree.h:16]: Thank you for pointing that out, I will come back to this comment when I'm merging the new branch then :)

File: src/polyhedralGravity/model/KDTree/KDTree.cpp (hunk 5 above)
Comment (Blauben) [reply src/polyhedralGravity/model/KDTree/KDTree.cpp:137]: For the operations std::set is defined for it's faster, like insertion and element fetching, However when iterating over all elements of the set, the red-black tree in the background of a std::set is significantly slower than a simple vector with array element access.

File: src/polyhedralGravity/model/KDTree/KDTree.cpp (hunk 3 above)
Comment (Blauben) [reply src/polyhedralGravity/model/KDTree/KDTree.cpp:113]: Resolved on a later branch. Plane required a custom constructor, leading to this syntax becoming invalid.

File: src/polyhedralGravity/model/KDTree/KDTree.cpp (hunk 2 above)
Comment (Blauben) [reply src/polyhedralGravity/model/KDTree/KDTree.cpp:64]: The vector is sorted by PlaneEvent . In practice this first compares the plane positions and should they be equal then only is the PlaneEventType considered. The outer while loop iterates over all planeEvents while the inner while loops iterate only over PlaneEvents that share the same plane and then differentiates between the PlaneEventTypes it encounters.

**Review by Blauben — COMMENTED** [review]
Completed review suggestions and addressed them

**Pull request merged**